### PR TITLE
feat(dup): protect private log from missing when duplication is enabled

### PR DESCRIPTION
**Pull request by neverchanje**
When duplication is enabled, we check the duplication progress each time log GC performs
to ensure log files won't be deleted while they are needed.

The pseudocode:

```cpp
        decree min_confirmed_decree = _duplication_mgr->min_confirmed_decree();
        decree cleanable_decree = _app->last_durable_decree();
        if (min_confirmed_decree >= 0 && min_confirmed_decree < cleanable_decree) {
            cleanable_decree = min_confirmed_decree;
        } else if (is_duplicating()) {
            return;
        }
```

### CODE REVIEW

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp [hunk 1]
```diff
@@ -80,5 +102,23 @@ void replica_duplicator_manager::remove_non_existed_duplications(
...
 
+void replica_duplicator_manager::set_confirmed_decree_non_primary(decree confirmed)
+{
+    // this function always runs in the same single thread with config-sync
+    if (_replica->status() == partition_status::PS_PRIMARY) {
+        return;
+    }
+
+    zauto_lock l(_lock);
+    remove_all_duplications();
```
Comment (hycdong) [inline src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 这里为什么要remove所有duplication呢？

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp [hunk 2]
```diff
@@ -80,5 +102,23 @@ void replica_duplicator_manager::remove_non_existed_duplications(
     }
 }
 
+void replica_duplicator_manager::set_confirmed_decree_non_primary(decree confirmed)
+{
+    // this function always runs in the same single thread with config-sync
+    if (_replica->status() == partition_status::PS_PRIMARY) {
```
Comment (hycdong) [inline src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 这个函数除了primary都有意义嘛？包括potential_secondary, secondary, inactive和error？

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp [hunk 3]
```diff
@@ -60,6 +60,28 @@ void replica_duplicator_manager::sync_duplication(const duplication_entry &ent)
     }
 }
 
+decree replica_duplicator_manager::min_confirmed_decree() const
```
Comment (hycdong) [inline src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: min_confirmed_decree在什么情况下会由非primary返回？

File: src/dist/replication/lib/replica_chkpt.cpp [hunk 4]
```diff
@@ -67,19 +69,44 @@ void replica::on_checkpoint_timer()
...
+            } else {
+                ddebug_replica("gc_private {}: min_confirmed_decree({}) last_durable_decree({})",
+                               enum_to_string(status()),
+                               min_confirmed_decree,
+                               last_durable_decree);
+            }
+        } else {
+            // protect the logs from being truncated
+            // if this app is in duplication
+            if (is_duplicating()) {
```
Comment (hycdong) [inline src/dist/replication/lib/replica_chkpt.cpp]: 这个直接返回的原因是什么？在热备过程中,min_confirmed_decree却是invalid_decree，在被设置成有效值之前，为了保护日志所以不gc？可不可以在这里加条debug日志解释一下呢？

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp (hunk 2 above)
Comment (neverchanje) [reply src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 通常来讲只对 secondary 有意义，因为 replica 处于其他状态的时候不会做 GC，就不需要用到 confirmed_decree。
但是这里为了保持简单，认为 non-primary 状态都可以。

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp (hunk 1 above)
Comment (neverchanje) [reply src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 避免一个 secondary 从 primary 调下来，它上面的热备份还在进行。

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp (hunk 2 above)
Comment (hycdong) [reply src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 那我建议把这个改成对secondary状态的检查，而不是非primary状态

File: src/dist/replication/lib/replica_check.cpp [hunk 5]
```diff
@@ -154,6 +159,9 @@ void replica::on_group_check(const group_check_request &request,
     } else if (is_same_ballot_status_change_allowed(status(), request.config.status)) {
         update_local_configuration(request.config, true);
     }
+    if (request.__isset.confirmed_decree) {
+        _duplication_mgr->set_confirmed_decree_non_primary(request.confirmed_decree);
```
Comment (hycdong) [inline src/dist/replication/lib/replica_check.cpp]: 如上所说，set_confirmed_decree_non_primary其实只对secondary有意义，除了secondary，group_check是允许inactive和potential_secondary状态的，这样就会给这些状态的replica也设置confirmed_decree，这个变量对这些状态其实是没意义的

File: src/dist/replication/lib/duplication/replica_duplicator_manager.cpp (hunk 2 above)
Comment (neverchanje) [reply src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: 这个改动有什么收益吗？我没看出改跟不改有什么区别，readability也没有提升？

Comment (hycdong) [reply src/dist/replication/lib/duplication/replica_duplicator_manager.cpp]: confirmed_decree既然只对secondary有意义，给其他状态的replica设置这个值的也没有必要额，为什么要放大设置值的范围呢？

File: src/dist/replication/lib/replica_chkpt.cpp (hunk 4 above)
Comment (neverchanje) [reply src/dist/replication/lib/replica_chkpt.cpp]: 已加日志

File: src/dist/replication/lib/duplication/test/replica_duplicator_test.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            // corner cases: next_status is INIT
+            duplicator->update_status_if_needed(duplication_status::DS_INIT);
+            ASSERT_EQ(duplicator->_status, duplication_status::DS_PAUSE);
+
+            duplicator->wait_all();
+        }
+    }
+};
+
+TEST_F(replica_duplicator_test, new_duplicator) { test_new_duplicator(); }
```
Comment (acelyc111) [inline src/dist/replication/lib/duplication/test/replica_duplicator_test.cpp]: 这种TEST里面转调一个函数的方式，是因为无法访问某些private变量吗？

Comment (neverchanje) [reply src/dist/replication/lib/duplication/test/replica_duplicator_test.cpp]: 是的
